Add label filter to subscriptions

scepticulous opened the pull request:
# Add label filters

## Description

As requested in #384 , many users want to filter their incoming PR- and issue events by certain labels. Based on the discussion in that issue and PR #797, the following syntax has been added.

```
# Notify events which have `todo` label
/github subscribe owner/repo +label:todo

# You can specify labels which have spaces when enclosed in quotes
/github subscribe owner/repo +label:"help wanted"

# Unsubscribe
/github unsubscribe owner/repo +label:"help wanted"
```

## Visuals

Rendering the list of labels:
![required-labels](https://user-images.githubusercontent.com/3408791/63572387-0a8e5b00-c583-11e9-9cac-6548fa334846.png)



## Key changes

The most significant change that was necessary, was changing the way we "parse" (un)subscribe commands. 

This was done by adding helper functions to parse the input and use them in the `Command` as well as in the `Subscription` model.
It was necessary to change both places because label parsing needs to be done early and feature evaluation is context-dependent based on subscribe/unsubscribe.

### Label parsing
The reason we have to do label parsing early is:

* Support for (quoted) whitespaces (Example: `label:"important feature")
* key-value pairs that can have the delimiter in its value. (Example: `label:priority:HIGH`)

### Feature evaluation

Combining parsed labels with parsed features during the command parsing phase was not possible, because feature values are context-dependent.

Example of context-depending booleans:
* `subscribe pulls` means `pulls: true`
* `unsubscribe pulls` means `pulls: false`

### Scope

To keep this PR "small" I only changed the data inside `Command` for `subscribe` and `unsubscribe`. This means, all other command should not be affected.

The only general change I made was replacing `command.args[0]` with `command.resource`

Example: `/github subscribe github/hub pulls issues`
* **subcommand** : subscribe
* **resource** : `github/hub`
* **args**: `pulls issues`


## PR origin
This PR is based on the prior work done in #797. Unfortunately that PR became out of sync and because of that, this branch is not associated with the original PR.
The contributions that @cou929 did, are correctly attributed.

## QA checklist

* [x] Subscribe works
  * [x] `/github subscribe +label:foo` command is accepted accepted
  * [x] `/github subscribe +label:foo` settings are stored correctly
  * [x] `/github subscribe +label:foo` settings are viewable
* [x] invalid labels have test coverage
* [x] invalid labels cause a proper error message
* [x] handle mixture of valid and invalid labels?
* [x] label presentation is pretty
* [x] filters have expected outcome
  * [x] incoming `issues` are discarded based on label filter
  * [x] incoming `issue comments` are discarded based on label filter
  * [x] incoming `PRs` are discarded based on label filter
* [x] Filtered content is logged
* [x] Unsubscribe works
  * [x] `/github unsubscribe +label:foo` command accepted
  * [x] `/github unsubscribe +label:foo` command persisted
  * [x] `/github unsubscribe +label:foo` command idempotent
* [x] Issues are accepted after removing filter
* [x] unsubscribe from invalid label does not cause side effects
* [x] removing filters have expected outcome
  * [x] incoming `issues` are no longer discarded
  * [x] incoming `issue comments` are no longer discarded
  * [x] incoming `PRs` are no longer discarded
* [x] (Optional) Shortcut for flushing all filters.
* [x] Document filter details in README
* [x] Add filters to help


Closes #384

izuzak commented:
Hey, @scepticulous :wave: Wow, this is super exciting! It would be great to be able to have more filtering capabilities like this one :+1: 

I wanted to share some thoughts/suggestions/questions I had -- looking forward to hearing what you think!

> ```
> # Notify events that have at least one `todo` or `help wanted` label
> # You can specify labels which have spaces when enclosed in quotes
> /github subscribe owner/repo label:todo label:"help wanted"
> ```

So, if I subscribe to `label:todo label:"help wanted"` then that means `todo OR "help wanted"`, not `todo AND "help wanted"`, correct?

If that's correct, then I'm wondering it this might be confusing for users since on the website -- the operator that's used when you list two labels is an AND, not an OR. For example, see this:

https://github.com/rails/rails/issues?utf8=%E2%9C%93&q=is%3Aissue+is%3Aopen+label%3A%22With+reproduction+steps%22+label%3Aactivesupport

So, searching for `label:"With reproduction steps" label:activesupport` gives you all issues with both labels, not either one. 

Also, I'm wondering if we could allow users to do both, i.e. to express an AND and an OR, so that they can fully customize what they want to receive. A syntax we might use for that might be:

```
label:foo,bar label:baz
```

which would be interpreted as: `(foo AND bar) OR baz`.

What do you think? 

Also, can you clarify which events exactly will be affected by the `label` filter? I'm guessing we will filter `issues` and `pulls` events based on the labels. But will we filter `comments` and `reviews` as well? So, if I do: `/github subscribe owner/repo reviews label:foo` will I only get reviews on pull requests which have the `foo` label on them? 

Lastly, I wanted to mention that GitHub webhooks have a `label` event itself:

https://developer.github.com/webhooks/#events

"label - Triggered when a repository's label is created, edited, or deleted."

With that in mind, I'm wondering if using `label` as a name might be confusing for users (who will exepct to get events related to label creations/updates/deletion) or be restrictive in case we in the future decide to add support to the Slack integration for these label events (because, how will we call that name if the name "label" is already taken)? 

Wondering if you have any thoughts about this and if this was already considered? Have we considered any alternatives here? For example, using something like `/github subscribe owner/repo issues#label:foo,bar` (now "label" is not a separate event you subscribe to, but a filter on issues, which in reality it really is since you're not subscribing to labels, but filtering other events based on labels).

scepticulous commented:
👋 @izuzak. Thanks for your great input. 

Let me try to answer your questions. Let me know if I missed something.


> If that's correct, then I'm wondering if this might be confusing for users since on the website -- the operator that's used when you list two labels is an AND, not an OR. For example, see this:

Yes, it is correct. Honestly, I don't know if it is confusing, I thought this would address most of the existing use-cases. However, I am not even sure if the 80% case is already satisfied with one filter. 

The core idea is to ship label filters and see if it solves most of the problems. I think if the push back on OR is too big, we can change it again. Generally speaking, the only way to really solve this is to add support for both, AND and OR.

> Also, I'm wondering if we could allow users to do both, i.e. to express an AND and an OR, so that they can fully customize what they want to receive. A syntax we might use for that might be:

Yes, we should do that. In this PR I only wanted to address the most basic case and it is already huge.  However, due to the relevance of this topic, I created #913 to discuss future filter syntax.


> Also, can you clarify which events exactly will be affected by the label filter?

Issues, issue comments and pull requests, but I will also check "PR comments". I added it to the README [here](https://github.com/integrations/slack/pull/911/files#diff-04c6e90faac2675aa89e2176d2eec7d8R135)


> Lastly, I wanted to mention that GitHub webhooks have a label event itself [...]  or be restrictive in case we in the future decide to add support to the Slack integration for these label events [...]

😭 Oh no, that is really bad.
TBH, I only used `label` instead of something like `label-filter`, to make it shorter, but the fact, that this causes a name conflict down the road (especially since label filters are stored as features) is really bad. I have to re-consider something like `label-filter`. This would pave the path to a future where filters are stored separately from features and we even consider `/github filter org/repo issue <filter expresssion>`

> Wondering if you have any thoughts about this and if this was already considered? Have we considered any alternatives here? For example, using something like `/github subscribe owner/repo issues#label:foo,bar`

This is interesting because it already enables event-specific filters (which is my long term goal anyway). I am just hesitant regarding the delimiter. We have to make sure that "#" is not used in values, or we just require quoting of delimiters, as CSV does.

izuzak commented:
> I think we don't really have matrics, on what is more important, so the pragmatic and lazy me things we should observe user feedback, even though I can see your point.
> ...
> Yes, we should do that. In this PR I only wanted to address the most basic case and it is already huge. However, due to the relevance of this topic, I created #913 to discuss future filter syntax.

Yeah, I agree that we should take an iterative approach if it's possible and there's no strong user demand for an OR+AND feature. My only concern is that we don't lead ourselves into a situation where we can't move from an OR to a OR+AND approach without breaking subscriptions or causing lots of confusion. For example, if today `label:foo label:bar` means `OR` and tomorrow it means `AND`, then that might both break existing subscriptions and confuse users who will run new subscribe commands using `label:foo label:bar` and expect an `OR` behavior. If there's a way for us to avoid such problems (e.g. by migrating existing subscriptions to ensure they still work as previously defined and messaging users when they subscribe with `label:foo label:bar` to explain what that will do) -- then maybe that would not irritate people too much? Maybe 😅 

> > Also, can you clarify which events exactly will be affected by the label filter?

> Issues, issue comments and pull requests, but I will also check "PR comments". I added it to the README here

Thanks for clarifying ❤️. I'm curious -- why wouldn't we filter PR reviews based on the label as well? What's the fundamental difference between an issue comment and a PR review here? Not saying we need to (and we can always add that later), just asking to understand since I'm guessing we'll get that question at some point from at least a few users.

> This is interesting because it already enables event-specific filters (which is my long term goal anyway). I am just hesitant regarding the delimiter. We have to make sure that "#" is not used in values, or we just require quoting of delimiters, as CSV does.

Yeah, it becomes a mess really fast, so if we can pick delimiters that offer a better user experience -- that's definitely 👍 👍 .

scepticulous commented:
> > Also, can you clarify which events exactly will be affected by the label filter?
> 
> Issues, issue comments and pull requests, but I will also check "PR comments". I added it to the README [here](https://github.com/integrations/slack/pull/911/files#diff-04c6e90faac2675aa89e2176d2eec7d8R135)

I double-checked and we are also filtering PR comments. I think this is what everybody expects and I just needed to test it. I updated the docs in https://github.com/integrations/slack/pull/911/commits/01c2ae45427b9c81a082859c4986e74ee4bcb9f4.


> My only concern is that we don't lead ourselves into a situation where we can't move from an OR to a OR+AND approach without breaking subscriptions or causing lots of confusion

💯 ! We will be able to migrate in case of a change, but I also don't want to change semantics. I will make a decision today and I think both defaults have good use-cases.

scepticulous commented:
## Update

After the great feedback that @izuzak provided, I decided to tweak this PR a little bit to make sure we are not blocking any future features and still move this over the finish line ASAP.

### Key changes

#### Filter syntax

The filter syntax will be changed to:
```
/github subscribe org/repo +label:review-needed
```
The `+` means "this is required". This opens the implementation for future features like `+assignee:xxx` or `+title:WIP`. 
In addition the `+` already indicates that we'll  **eventually** have  `-label:foo` to ignore/blacklist a label or other attribute.

This PR **only** introduces `+label` and we will see where we will go from there.
I hope everybody else likes this super concise and short syntax too. I ❤️ it 🎉 

#### AND and OR

I decided to drop support for multiple labels in this PR. The goal was always to ship something that solves an important use-case and iterate from there.
This means only the last `+label` will be considered and adding it repeatedly will always override/update the previous value.


Why:

The question around AND and OR is complex and I want to address @izuzak's concerns of making sure the syntax is stable and not confusing, but postponing this capability.
I think OR is the most common use-case but I want to first make sure I am happy with the combinators before pushing it to production and changing it again.

`cc` @wilhelmklopp @jnraine @jules2689

izuzak commented:
Thanks for the great updates, @scepticulous ⚡️ 

> I double-checked and we are also filtering PR comments. I think this is what everybody expects and I just needed to test it. I updated the docs in 01c2ae4.

Just to clarify, I'm asking about the `reviews` events:

<img width="1377" alt="Screenshot 2019-08-23 at 09 18 33" src="https://user-images.githubusercontent.com/38924/63574416-0284ea00-c588-11e9-89ec-8cb845104b51.png">

So, why are we filtering `issues`, `pulls` and `comments`, but not `reviews`? What's the difference between filtering `comments` and `reviews`? 

> I decided to drop support for multiple labels in this PR. The goal was always to ship something that solves an important use-case and iterate from there.

I think this is a great first step 👏 

> The + means "this is required". This opens the implementation for future features like +assignee:xxx or +title:WIP.
> In addition the + already indicates that we'll eventually have -label:foo to ignore/blacklist a label or other attribute.

I like this syntax as well 👏

scepticulous commented:
> So, why are we filtering `issues`, `pulls` and `comments`, but not `reviews`? What's the difference between filtering `comments` and `reviews`?
> 
> >

@izuzak You are the best 🎉 🥇 🏆 

Its the second time you spotted incorrect/incomplete documentation. The implementation does filter review events already. I fixed and improved he docs WRT filtered events in https://github.com/integrations/slack/pull/911/commits/b717bc0dffe311abdcf4f1fd6bd0d3ef22cc190b

wilhelmklopp commented on lib/models/subscription.js:
```diff
@@ -112,34 +185,122 @@ module.exports = (sequelize, DataTypes) => {
+    // visibility public
+    // update the subscription by disabling a set of features
+    disable(settings) {
+      if (settings) {
```
Should this throw if settings is falsy?

wilhelmklopp commented on test/integration/__snapshots__/help.test.js.snap:
```diff
@@ -44,6 +44,20 @@ Object {
+      "mrkdwn_in": Array [
+        "text",
+      ],
+      "text": "Create required-label. Issues, Comments, PRs without that label will be ignored.:
```
Do the `.` and `:` together make this look a bit weird?

wilhelmklopp commented on test/integration/notifications.test.js:
```diff
@@ -1256,5 +1259,92 @@ describe('Integration: notifications', () => {
+        name: 'issues',
+        payload: issuePayload,
+      });
+    });
```
Tests look greaaaat 💪

wilhelmklopp commented on test/integration/subscriptions.test.js:
```diff
@@ -326,6 +327,69 @@ describe('Integration: subscriptions', () => {
+
+        await request.post('/slack/command').use(slackbot)
+          .send(fixtures.slack.command({
+            text: 'subscribe bkeepers/dotenv +label:"help wanted" +label:todo +label:priority:MUST +label:\'good first issue\'',
```
Are we still allowing multiple label filters for now?

scepticulous replied:
For some weird reasons the tests are passing even though we do not allow that anymore.
I have to take a look, why the integration test was passing, despite my changes 🤔

scepticulous replied:
Fixed in https://github.com/integrations/slack/pull/911/commits/dc4fe0f17be107a46b63e0306caf5c18a3c3e9d9

scepticulous commented on lib/activity/filters.js:
```diff
@@ -0,0 +1,23 @@
+
+  if (Array.isArray(subscription.settings.label) && subscription.settings.label.length === 0) {
+    return false;
+  }
```
Do you have an example for a more implicit syntax for the `empty array` check?
I like the idea of truthiness and falseyness, but calling `length` on something that is not an array would raise an exception, wouldn't it?

scepticulous commented on lib/activity/router.js:
```diff
@@ -76,9 +77,23 @@ module.exports = function route(callback) {
+        const issue = context.payload.issue || context.payload.pull_request;
+        if (shouldFilterByLabel(subscription, issue)) {
+          const labels = issue.labels.map(l => l.name);
+          const whitelist = subscription.settings.label;
```
It's more a list of **required** and not really **allowed** values and this maps 1:1 to the idea of a whitelist. 
we could also just discard the temporary var and use the `required_labels` property itself if that reduces confusion.

scepticulous commented on lib/activity/router.js:
```diff
@@ -76,9 +77,23 @@ module.exports = function route(callback) {
+        // Label filtering
+        const issue = context.payload.issue || context.payload.pull_request;
+        if (shouldFilterByLabel(subscription, issue)) {
+          const labels = issue.labels.map(l => l.name);
```
I agree. This was just a leftover. I changed it.

scepticulous commented on lib/slack/command.js:
```diff
@@ -11,14 +12,22 @@ module.exports = class Command {
+      // command text without subcommand
+      this.text = argString;
+
+      // be smart about subscription target
```
Haha 😆 . That is a leftover. I added the comment before implementing the `resource` change.

scepticulous commented on lib/settings-helper.js:
```diff
@@ -0,0 +1,116 @@
+}
+
+// ['comments:all', 'issues'] => {comments: all, issues: true/false}
+function parseFeatures(features, booleanContext) {
```
For `subscribe` (true context) we want to convert like this
```javascript
['pulls', issues'] => {pulls: true, issues: true}
```
But for `unsubscribe` (false context), we want to do:
```javascrript
['pulls', issues'] => {pulls: false, issues: false}
```

scepticulous commented on lib/settings-helper.js:
```diff
@@ -0,0 +1,116 @@
+
+function extractLabelValue(arg) {
+  const match = labelPattern().exec(arg);
+  return match[1] || match[2];
```
This is confusing to me too... When the second match group matches (quoted strings), then m[1] is undefined.

```javascript
> const p = /(?:\+label:(?:([A-Za-z0-9-:./]+)|["']([A-Za-z0-9-:./\s]+)["']) ?)+$/g;
undefined
> let m = p.exec('+label:"help wanted"')
undefined
> m[0]
'+label:"help wanted"'
> m[1]
undefined
> m[2]
'help wanted
```

jfragoulis commented:
Hello

Documentation of the feature is missing from https://slack.github.com and https://slack.com/apps/A8GBNUWU8-github

DDwibedy commented:
hi @dennissivia I am facing another issue now, I have subscribed my slack channel to get notified if a specific label is tagged to my PR . I am getting notification also but the issue is it's also sending the notifications where the specific label is not tagged. And it spamming my channel with irrelevant commit details in my channel.

Below command executed :
/github subscribe owner/repo-name +label:white-label

Note: My label name is "white-label" . Sorry I can't put my repo details here in the thread.

Please suggest!
